File: packages/totem-sdk/packages/proof/src/anchor.ts
/**
 * Anchor helpers.
 *
 * createAnchorCommitment: deterministic hash of a proof for on-chain anchoring.
 * attachAnchor: attach an AnchorRef to a SignedProof without changing proofId.
 * verifyAnchorRef: recompute commitment and compare.
 */

import { sha3_256 } from '@noble/hashes/sha3.js';
import { toHex, canonicalJson } from './canonical.js';
import type { SignedProof, AnchorRef } from './types.js';

/**
 * Compute a deterministic anchor commitment for a SignedProof.
 * Used as the hash submitted to an anchoring provider (e.g. Integritas).
 */
export function createAnchorCommitment(signedProof: SignedProof): string {
  const input = 'totem-anchor' + signedProof.proofId + canonicalJson(signedProof.subject);
  return toHex(sha3_256(new TextEncoder().encode(input)));
}

/**
 * Attach an AnchorRef to a SignedProof.
 * The proofId MUST remain unchanged — anchor is mutable metadata outside the signed region.
 */
export function attachAnchor(signedProof: SignedProof, anchorRef: AnchorRef): SignedProof {
  return { ...signedProof, anchor: anchorRef };
}

/**
 * Verify that an AnchorRef.hash matches the expected commitment for this proof.
 */
export function verifyAnchorRef(signedProof: SignedProof, anchorRef: AnchorRef): boolean {
  const expected = createAnchorCommitment(signedProof);
  return expected === anchorRef.hash;
}

File: packages/totem-sdk/packages/proof/rust/src/anchor.rs
use sha3::{Digest, Sha3_256};

use crate::canonical::{canonical_json, to_hex};
use crate::types::SignedProof;

pub fn create_anchor_commitment(signed_proof: &SignedProof) -> String {
    let subject_json = serde_json::to_value(&signed_proof.subject).unwrap_or_default();
    let input = format!(
        "totem-anchor{}{}",
        signed_proof.proof_id,
        canonical_json(&subject_json)
    );
    let mut hasher = Sha3_256::new();
    hasher.update(input.as_bytes());
    to_hex(&hasher.finalize())
}

pub fn verify_anchor_ref(signed_proof: &SignedProof, anchor_hash: &str) -> bool {
    let expected = create_anchor_commitment(signed_proof);
    expected == anchor_hash
}
